### PR TITLE
fix TypedArray set

### PR DESCRIPTION
**Pull request by sablib**
It should follow the spec and behave the same with TypedArray in Typescript.

Re: #

### Changelog

*

-------

### Continuous Integration

This pull request:

* [ ] needs automatic test cases check.
  > Manual trigger with `@cocos-robot run test cases` afterward.
* [ ] does not change any runtime related code or build configuration
  > If any reviewer thinks the CI checks are needed, please uncheck this option, then close and reopen the issue.

-------

### Compatibility Check

This pull request:

* [ ] changes public API, and have ensured backward compatibility with [deprecated features](https://github.com/cocos/cocos-engine/blob/v3.5.0/docs/contribution/deprecated-features.md).
* [ ] affects platform compatibility, e.g. system version, browser version, platform sdk version, platform toolchain, language version, hardware compatibility etc.
* [ ] affects file structure of the build package or build configuration which requires user project upgrade.
* [ ] **introduces breaking changes**, please list all changes, affected features and the scope of violation.

### CODE REVIEW

**Review by minggo — APPROVED** [review]

File: native/cocos/core/TypedArray.h [hunk 1]
```diff
@@ -192,9 +192,9 @@ class TypedArrayTemp {
     }
 
     void set(ArrayBuffer *buffer, uint32_t offset) {
-        CC_ASSERT(buffer->byteLength() + offset <= _byteEndPos);
+        CC_ASSERT(buffer->byteLength() + _byteOffset + offset * BYTES_PER_ELEMENT <= _byteEndPos);
         CC_ASSERT(_buffer);
-        memcpy(_buffer->_data + offset, buffer->_data, buffer->byteLength());
+        memcpy(_buffer->_data + _byteOffset + offset * BYTES_PER_ELEMENT, buffer->_data, buffer->byteLength());
```
Comment (dumganhar) [inline native/cocos/core/TypedArray.h:197]: Good catch.
In this case, I think another two set methods should also consider the current `_byteOffset` while copying data.
```cpp
template <typename T>
template <typename SrcType>
typename std::enable_if_t<std::is_same<T, SrcType>::value, void> TypedArrayTemp<T>::set(const TypedArrayTemp<SrcType> &array, uint32_t offset) {
    CC_ASSERT(_buffer);
    uint32_t dstByteOffset = offset * BYTES_PER_ELEMENT;
    uint32_t srcByteOffset = array.byteOffset();
    uint32_t srcCount = array.length();
    CC_ASSERT(dstByteOffset + srcCount * TypedArrayTemp<SrcType>::BYTES_PER_ELEMENT <= _byteEndPos);
    memcpy(_buffer->_data + dstByteOffset, array._buffer->_data + srcByteOffset, array.byteLength());
}
template <typename T>
template <typename SrcType>
typename std::enable_if_t<!std::is_same<T, SrcType>::value, void> TypedArrayTemp<T>::set(const TypedArrayTemp<SrcType> &array, uint32_t offset) {
    CC_ASSERT(_buffer);
    uint32_t dstByteOffset = offset * BYTES_PER_ELEMENT;
    uint32_t srcByteOffset = array.byteOffset();
    uint32_t srcCount = array.length();
    uint32_t remainCount = (_byteEndPos - dstByteOffset) / BYTES_PER_ELEMENT;
    CC_ASSERT_LE(srcCount, remainCount);
    for (uint32_t i = 0; i < srcCount; ++i) {
        (*this)[offset + i] = reinterpret_cast<T>(array[i]);
    }
}

```

Comment (sablib) [reply native/cocos/core/TypedArray.h:197]: Add a commit to fix these function templates.

**Comment by dumganhar** [comment]
@cocos-robot run test cases

**Review by dumganhar — APPROVED** [review]

**Comment by dumganhar** [comment]
@cocos-robot run test cases

**Comment by dumganhar** [comment]
@cocos-robot run test cases

**Pull request merged**